### PR TITLE
[SYCL] Fix id equal operator

### PR DESCRIPTION
**Pull request by idubinov**
Signed-off-by: Igor Dubinov <igor.dubinov@intel.com>

### CODE REVIEW

File: sycl/include/CL/sycl/id.hpp [hunk 1]
```diff
@@ -177,33 +192,40 @@ __SYCL_INLINE namespace cl {
...
+// OP is: ==, !=
+#define __SYCL_GEN_OPT(op)                                                     \
+  bool operator op(id<1> &rhs) const {                                         \
+    return this->common_array[0] op rhs.common_array[0];                       \
+  }                                                                            \
+  bool operator op(size_t &rhs) const { return this->common_array[0] op rhs; } \
+  bool operator op(range<1> &rhs) const {                                      \
+    return this->common_array[0] op rhs[0];                                    \
+  }                                                                            \
+  /* The equal check of size_t op id<1> goes thougth id<1> -> size_t           \
```
Comment (AlexeySachkov) [inline sycl/include/CL/sycl/id.hpp]:
```suggestion
  /* The equality check of size_t op id<1> goes thougth id<1> -> size_t           \
```

**Comment by rolandschulz** [comment]
Ruslan figured out a solution which avoids the template specialization. Instead of using the template specialization we can disable the conversion operator by using 
```c++
operator typename std::conditional<dimensions == 1, size_t, void>::type() const
```
Should we undo the template specialization from the previous PR? Then this fix isn't needed.

**Comment by idubinov** [comment]
@rolandschulz , the current fix is still needed but only for id and size_t operands. In case of id to size_t conversion there is ambiguity in operators == and != : 

id == size_t : id can be casted to size_t using implicit conversion (size_t==size_t) or both can be casted to array using array(size_t) constructor (array==array operator ). 

To avoid this ambiguity i decided to overload equal operators for id and integral types.

I am planning to prepare refactoring with Ruslan's suggestion as another PR

**Comment by MrSidims** [comment]
> @rolandschulz , the current fix is still needed but only for id and size_t operands. In case of id to size_t conversion there is ambiguity in operators == and != :
> 
> id == size_t : id can be casted to size_t using implicit conversion (size_t==size_t) or both can be casted to array using array(size_t) constructor (array==array operator ).
> 
> To avoid this ambiguity i decided to overload equal operators for id and integral types.
> 
> I am planning to prepare refactoring with Ruslan's suggestion as another PR

Please put a problem description in the commit message.

**Comment by MrSidims** [comment]
The patch itself LGTM

**Pull request closed without merging**